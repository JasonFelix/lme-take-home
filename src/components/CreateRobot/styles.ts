import styled from "styled-components";

const Title = styled.div`
  color: var(--primary-text);
  font-size: var(--header-text);
  margin-bottom: var(--margin-m);
`;

const Description = styled.div`
  color: var(--primary-text);
  font-size: 12px;
  margin-bottom: var(--margin-m);
`;

const DimensionsInput = styled.div`
  display: flex;
  flex-direction: row;
`;

const Header = styled.div`
  font-size: 14px;
  color: var(--primary-text);
  position: relative;
`;

const HorizontalGroup = styled.div`
  flex: 1;
  display: flex;
  flex-direction: row;
`;

const Group = styled.div`
  flex: 1;
`;

const Text = styled.div`
  font-size: 12px;
  color: var(--primary-text);
  margin-top: var(--margin-m);
`;

export {
  Title,
  Description,
  DimensionsInput,
  Header,
  HorizontalGroup,
  Group,
  Text,
};
