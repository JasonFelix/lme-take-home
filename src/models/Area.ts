import { makeAutoObservable } from "mobx";
import Robot from "./Robot";
import { Coordinates, Dimensions } from "./types";

export default class Area {
  _robots: Array<Robot> = [];

  _scents: Map<string, boolean> = new Map();

  /**
   *
   * @param width : Size of grid width. Default size is 50.
   * @param height : Size of grid width. Default size is 50.
   */
  constructor(
    private _dimensions: Dimensions,
    private _maxDimensions: Dimensions = { width: 50, height: 50 }
  ) {
    if (
      this.dimensions.width > this._maxDimensions.width ||
      this._maxDimensions.height > 50
    ) {
      throw new Error(
        `Grid height / width too large. The max width is ${this._maxDimensions.width} / height is ${this._maxDimensions.height}.`
      );
    }
    makeAutoObservable(this);
  }

  addRobot(robot: Robot): void {
    robot.area = this;

    while (robot.hasInstructions() && !robot.isLost()) {
      const prev = robot.coordinates;
      robot.move();
      const post = robot.coordinates;
      if (!this.withinBounds(post)) {
        this.scents.set(JSON.stringify(prev), true);
        // eslint-disable-next-line no-continue
        continue;
      }
    }

    this.robots.push(robot);
  }

  get robots(): Robot[] {
    return this._robots;
  }

  set robots(robots: Robot[]) {
    this._robots = [...robots];
  }

  get scents(): Map<string, boolean> {
    return this._scents;
  }

  set scents(scents: Map<string, boolean>) {
    this._scents = { ...scents };
  }

  get dimensions(): Dimensions {
    return this._dimensions;
  }

  set dimensions({ width, height }: Dimensions) {
    if (width > this._maxDimensions.width || this._maxDimensions.height > 50) {
      throw new Error(
        `Grid height / width too large. The max width is ${this._maxDimensions.width} / height is ${this._maxDimensions.height}.`
      );
    }
    this.dimensions.width = width;
    this.dimensions.height = height;
  }

  purge(): void {
    this.robots = [];
    this.scents = new Map();
    this.dimensions.width = 0;
    this.dimensions.height = 0;
  }

  withinBounds({ x, y }: Coordinates): boolean {
    return (
      x >= 0 &&
      y >= 0 &&
      x <= this.dimensions.width &&
      y <= this.dimensions.height
    );
  }
}
