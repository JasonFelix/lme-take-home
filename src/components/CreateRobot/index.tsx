import { Input, Select, Space } from 'antd';

import NumberInput from "../NumberInput";
import { ReactComponent as RobotSVG } from '../../images/robot.svg';
import { Button } from 'antd';
import { useState } from "react";
import InstructionSelect from "../InstructionSelect";
import { Coordinates, Instruction, Orientation } from "../../models/types";
import Area from "../../models/Area";
import { observer } from "mobx-react";
import Robot from "../../models/Robot";
import { Title, Text, HorizontalGroup, Group, Header, DimensionsInput } from "./styles";
import { Container, LeftContainer, RightContainer } from '../Panel/styles';


interface Props {
  area: Area;
  onActivate: () => void;
}

const CreateRobot = observer(({ area, onActivate }: Props) => {
  const [instructions, setInstructions] = useState<Array<Instruction>>([]);
  const [coordinates, setCoordinates] = useState<Coordinates>({ x: 0, y: 0 });
  const [orientation, setOrientation] = useState<Orientation>('N');

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const uppercase = e.target.value.toUpperCase();
    const filtered = uppercase.split('').filter(c => ['L', 'R', 'F'].includes(c));
    setInstructions(filtered.slice(0, 100) as Instruction[]);
  }

  const reset = () => {
    setInstructions([]);
    setCoordinates({ x: 0, y: 0 });
    setOrientation('N');
  }

  const activateRobot = () => {
    const robot = new Robot(coordinates, orientation, instructions);
    area.addRobot(robot);
    onActivate();
    reset();
  }

  const addInstruction = (instruction: Instruction) => {
    setInstructions([...instructions, instruction].slice(0, 100) as Instruction[]);
  }

  const removeInstruction = (index: number) => {
    setInstructions([...instructions.slice(0, index), ...instructions.slice(index + 1)]);
  }

  const instructionsString = instructions.join('');

  return <Container>
    <LeftContainer>
      <RobotSVG />
    </LeftContainer>
    <RightContainer>
      <Space direction="vertical" style={{width: '100%', paddingRight: '16px'}}>
        <Title>Create a Robot</Title>
        <HorizontalGroup>
          <Group>
            <Header>Starting Coordinates</Header>
            <DimensionsInput>
              <NumberInput label="X" value={coordinates.x} min={0} max={50} callback={x => { setCoordinates({ ...coordinates, x }) }} />
              <NumberInput label="Y" value={coordinates.y} min={0} max={50} callback={y => { setCoordinates({ ...coordinates, y }) }} />
            </DimensionsInput>
          </Group>
          <Group>
            <Header>Starting Orientation</Header>
            <DimensionsInput>
              <Select defaultValue="N" value={orientation} onChange={(orientation: Orientation) => setOrientation(orientation)}>
                <Select.Option value="N">North</Select.Option>
                <Select.Option value="S">South</Select.Option>
                <Select.Option value="E">East</Select.Option>
                <Select.Option value="W">West</Select.Option>
              </Select>
            </DimensionsInput>
          </Group>
        </HorizontalGroup>
        <Header>Search Instructions</Header>
        <Input type="text" value={instructionsString} onChange={onChange} />
        <HorizontalGroup>
          <Space direction="horizontal">
            <Button onClick={() => addInstruction('L')}>Turn Left</Button>
            <Button onClick={() => addInstruction('F')}>Walk Forward</Button>
            <Button onClick={() => addInstruction('R')}>Turn Right</Button>
          </Space>
        </HorizontalGroup>
        <InstructionSelect instructions={instructions} callback={removeInstruction} />
        <Text>Instructions {instructions.length} / 100</Text>
        <Button type="primary" onClick={activateRobot}>Activate Robot</Button>
      </Space>
    </RightContainer>
  </Container>
});

export default CreateRobot;