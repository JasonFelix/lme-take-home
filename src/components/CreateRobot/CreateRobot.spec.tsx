import renderer from "react-test-renderer";
import Area from "../../models/Area";
import CreateRobot from ".";

it("CreateRobot renders correctly", () => {
  const area: Area = new Area(
    { width: 0, height: 0 },
    { width: 50, height: 50 }
  );
  const tree = renderer
    .create(<CreateRobot area={area} onActivate={() => ({})} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
