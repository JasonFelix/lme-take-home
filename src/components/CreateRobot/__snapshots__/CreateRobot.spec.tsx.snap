// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRobot renders correctly 1`] = `
<div
  className="sc-crrszt cPETOR"
>
  <div
    className="sc-dQoVA ehogfY"
  >
    <svg>
      robot.svg
    </svg>
  </div>
  <div
    className="sc-bqyKOL bnILcz"
  >
    <div
      className="ant-space ant-space-vertical"
      style={
        Object {
          "paddingRight": "16px",
          "width": "100%",
        }
      }
    >
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-gKseQn NflxS"
        >
          Create a Robot
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-jrAFXE kidoSs"
        >
          <div
            className="sc-kEjbQP jvxEBd"
          >
            <div
              className="sc-pGacB eEmGhR"
            >
              Starting Coordinates
            </div>
            <div
              className="sc-fubCzh jirHnz"
            >
              <div
                className="sc-bdfBQB fufubI"
                onMouseDown={[Function]}
              >
                <span
                  className="sc-gsTEea enATLc"
                >
                  X
                </span>
                <input
                  className="sc-dlfnuX haFIGb"
                  max={50}
                  min={0}
                  onChange={[Function]}
                  type="number"
                  value={0}
                />
              </div>
              <div
                className="sc-bdfBQB fufubI"
                onMouseDown={[Function]}
              >
                <span
                  className="sc-gsTEea enATLc"
                >
                  Y
                </span>
                <input
                  className="sc-dlfnuX haFIGb"
                  max={50}
                  min={0}
                  onChange={[Function]}
                  type="number"
                  value={0}
                />
              </div>
            </div>
          </div>
          <div
            className="sc-kEjbQP jvxEBd"
          >
            <div
              className="sc-pGacB eEmGhR"
            >
              Starting Orientation
            </div>
            <div
              className="sc-fubCzh jirHnz"
            >
              <div
                className="ant-select ant-select-single ant-select-show-arrow"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
              >
                <div
                  className="ant-select-selector"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <span
                    className="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autoComplete="off"
                      className="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      readOnly={true}
                      role="combobox"
                      style={
                        Object {
                          "opacity": 0,
                        }
                      }
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    className="ant-select-selection-item"
                    title="North"
                  >
                    North
                  </span>
                </div>
                <span
                  aria-hidden={true}
                  className="ant-select-arrow"
                  onMouseDown={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    className="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-pGacB eEmGhR"
        >
          Search Instructions
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <input
          className="ant-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-jrAFXE kidoSs"
        >
          <div
            className="ant-space ant-space-horizontal ant-space-align-center"
            style={Object {}}
          >
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginRight": 8,
                }
              }
            >
              <button
                className="ant-btn"
                onClick={[Function]}
                type="button"
              >
                <span>
                  Turn Left
                </span>
              </button>
            </div>
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginRight": 8,
                }
              }
            >
              <button
                className="ant-btn"
                onClick={[Function]}
                type="button"
              >
                <span>
                  Walk Forward
                </span>
              </button>
            </div>
            <div
              className="ant-space-item"
              style={Object {}}
            >
              <button
                className="ant-btn"
                onClick={[Function]}
                type="button"
              >
                <span>
                  Turn Right
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-hKgJUU irkFgh"
        >
          <div
            className="ant-empty"
          >
            <div
              className="ant-empty-image"
            >
              <svg
                className="ant-empty-img-default"
                height="152"
                viewBox="0 0 184 152"
                width="184"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                >
                  <g
                    transform="translate(24 31.67)"
                  >
                    <ellipse
                      className="ant-empty-img-default-ellipse"
                      cx="67.797"
                      cy="106.89"
                      rx="67.797"
                      ry="12.668"
                    />
                    <path
                      className="ant-empty-img-default-path-1"
                      d="M122.034 69.674L98.109 40.229c-1.148-1.386-2.826-2.225-4.593-2.225h-51.44c-1.766 0-3.444.839-4.592 2.225L13.56 69.674v15.383h108.475V69.674z"
                    />
                    <path
                      className="ant-empty-img-default-path-2"
                      d="M101.537 86.214L80.63 61.102c-1.001-1.207-2.507-1.867-4.048-1.867H31.724c-1.54 0-3.047.66-4.048 1.867L6.769 86.214v13.792h94.768V86.214z"
                      transform="translate(13.56)"
                    />
                    <path
                      className="ant-empty-img-default-path-3"
                      d="M33.83 0h67.933a4 4 0 0 1 4 4v93.344a4 4 0 0 1-4 4H33.83a4 4 0 0 1-4-4V4a4 4 0 0 1 4-4z"
                    />
                    <path
                      className="ant-empty-img-default-path-4"
                      d="M42.678 9.953h50.237a2 2 0 0 1 2 2V36.91a2 2 0 0 1-2 2H42.678a2 2 0 0 1-2-2V11.953a2 2 0 0 1 2-2zM42.94 49.767h49.713a2.262 2.262 0 1 1 0 4.524H42.94a2.262 2.262 0 0 1 0-4.524zM42.94 61.53h49.713a2.262 2.262 0 1 1 0 4.525H42.94a2.262 2.262 0 0 1 0-4.525zM121.813 105.032c-.775 3.071-3.497 5.36-6.735 5.36H20.515c-3.238 0-5.96-2.29-6.734-5.36a7.309 7.309 0 0 1-.222-1.79V69.675h26.318c2.907 0 5.25 2.448 5.25 5.42v.04c0 2.971 2.37 5.37 5.277 5.37h34.785c2.907 0 5.277-2.421 5.277-5.393V75.1c0-2.972 2.343-5.426 5.25-5.426h26.318v33.569c0 .617-.077 1.216-.221 1.789z"
                    />
                  </g>
                  <path
                    className="ant-empty-img-default-path-5"
                    d="M149.121 33.292l-6.83 2.65a1 1 0 0 1-1.317-1.23l1.937-6.207c-2.589-2.944-4.109-6.534-4.109-10.408C138.802 8.102 148.92 0 161.402 0 173.881 0 184 8.102 184 18.097c0 9.995-10.118 18.097-22.599 18.097-4.528 0-8.744-1.066-12.28-2.902z"
                  />
                  <g
                    className="ant-empty-img-default-g"
                    transform="translate(149.65 15.383)"
                  >
                    <ellipse
                      cx="20.654"
                      cy="3.167"
                      rx="2.849"
                      ry="2.815"
                    />
                    <path
                      d="M5.698 5.63H0L2.898.704zM9.259.704h4.985V5.63H9.259z"
                    />
                  </g>
                </g>
              </svg>
            </div>
            <div
              className="ant-empty-description"
            >
              <span>
                No instructions
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-iqHYmW cxSVAc"
        >
          Instructions
           
          0
           
          / 100
        </div>
      </div>
      <div
        className="ant-space-item"
        style={Object {}}
      >
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            Activate Robot
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
